lib/logstash/filters/aggregate.rb

```diff
@@ -0,0 +1,228 @@
+# How it works :
+# - the filter needs a "task_id" to correlate events (log lines) of a same task
+# - at the task beggining, filter creates a map, attached to task_id
+# - for each event, you can execute code using 'event' and 'map' (for instance, copy an event field to map)
+# - in the final event, you can execute a last code (for instance, add map data to final event)
+# - after the final event, the map attached to task is deleted
+# - in one filter configuration, it is recommanded to define a timeout option to protect the feature against unterminated tasks. It tells the filter to delete expired maps
+# - if no timeout is defined, by default, all maps older than 1800 seconds are automatically deleted
+#
+#
+class LogStash::Filters::Aggregate < LogStash::Filters::Base
+
+	config_name "aggregate"
+
+	# The expression defining task ID to correlate logs. +
+	# This value must uniquely identify the task in the system +
+	# Example value : "%{application}%{my_task_id}" +
+	config :task_id, :validate => :string, :required => true
+
+	# The code to execute to update map, using current event. +
+	# Or on the contrary, the code to execute to update event, using current map. +
+	# You will have a 'map' variable and an 'event' variable available (that is the event itself). +
+	# Example value : "map['dao_duration'] += event['duration']" +
+	config :code, :validate => :string, :required => true
+
+	# Tell the filter what to do with aggregate map (default :  "create_or_update"). +
+	# create: create the map, and execute the code only if map wasn't created before +
+	# update: doesn't create the map, and execute the code only if map was created before +
+	# create_or_update: create the map if it wasn't created before, execute the code in all cases +
+	config :map_action, :validate => :string, :default => "create_or_update"
+
+	# Tell the filter that task is ended, and therefore, to delete map after code execution.  
+	config :end_of_task, :validate => :boolean, :default => false
+
+	# The amount of seconds after a task "end event" can be considered lost. +
+	# The task "map" is evicted. +
+	# The default value is 0, which means no timeout so no auto eviction. +
+	config :timeout, :validate => :number, :required => false, :default => 0
+
+	
+	# Default timeout (in seconds) when not defined in plugin configuration
+	DEFAULT_TIMEOUT = 1800
+
+	# This is the state of the filter.
+	# For each entry, key is "task_id" and value is a map freely updatable by 'code' config
+	@@aggregate_maps = {}
+
+	# Mutex used to synchronize access to 'aggregate_maps'
+	@@mutex = Mutex.new
+
+	# Aggregate instance which will evict all zombie Aggregate elements (older than timeout)
+	@@eviction_instance = nil
+
+	# last time where eviction was launched
+	@@last_eviction_timestamp = nil
+
+	# Initialize plugin
+	public
+	def register
+		# process lambda expression to call in each filter call
+		eval("@codeblock = lambda { |event, map| #{@code} }", binding, "(aggregate filter code)")
```

Comment:
It is not less secure than `ruby` filter plugin.
Then, I don't see another way to do what `aggregate` plugin do.
And to my mind, it is the only way to benefit all the power of the `aggregate` plugin.
As you can imagine, I use this plugin for my own need (in production), and because of that `code` option, the plugin is clearly powerfull.